polkadot/node/network/bridge/src/network.rs

```diff
@@ -45,51 +48,50 @@ const LOG_TARGET: &'static str = "parachain::network-bridge-net";
 /// messages that are compatible with the passed peer set, as that is currently not enforced by
 /// this function. These are messages of type `WireMessage` parameterized on the matching type.
 pub(crate) fn send_message<M>(
-	net: &mut impl Network,
 	mut peers: Vec<PeerId>,
 	peer_set: PeerSet,
 	version: ProtocolVersion,
-	protocol_names: &PeerSetProtocolNames,
 	message: M,
 	metrics: &super::Metrics,
+	network_notification_sinks: &Arc<Mutex<HashMap<(PeerSet, PeerId), Box<dyn MessageSink>>>>,
 ) where
 	M: Encode + Clone,
 {
 	if peers.is_empty() {
 		return
 	}
+
 	let message = {
 		let encoded = message.encode();
 		metrics.on_notification_sent(peer_set, version, encoded.len(), peers.len());
 		encoded
 	};
 
+	let notification_sinks = network_notification_sinks.lock();
```

Comment:
Maybe it will be more useful to obtain this lock merely when we get a concrete `sink` to send a message to? That would lead to more locks/unlocks but it might improve the concurrency if this lock is contended by other users.

Reply:
I'm not sure I understand what you're suggesting, could you be more specific?

I don't see how we could get rid of this lock on the `HashMap` since `NetworkBridgeRx` can modify it at any time by adding/removing peers.

Reply:
I mean to lock it only when getting information. It might also a good point to use `rwlock` instead of `mutex`, as we have much more reads than modify ops.